Add minimal runner to open a map in play mode.

Nemikolh opened the pull request:
@Vaelden : Could you push on this branch the stuff you showed me yesterday? Don't worry if you are unsure of where to put it, I will move it afterwards. I just want you to get the credit for your work. :)

Nemikolh committed:
Add minimal runner to open a map in play mode.

Nemikolh committed:
Upgrade webpack-dev-server and add PhysicsEngine.

Vaelden commented on editor/modules/runner/physics-engine.ts:
```diff
@@ -0,0 +1,72 @@
+    update(dt: number) {
+        // Update speed.
+        for (let entity of this._entities) {
+            let dir =  direction(entity.next_pos, entity.pos);
```
Origin / Destination look reversed here

Nemikolh replied:
Oops! That's corret.

Vaelden commented on editor/modules/runner/physics-engine.ts:
```diff
@@ -0,0 +1,72 @@
+        let diff;
+        for (let entity of this._entities) {
+
+            diff = entity.next_pos.x - entity.pos.x;
```
I assume `next_pos` will hold the updates coming from the server, and will try to do some prediction if non is received, correct?

But after the first iteration, `entity.pos` is assigned to `entity.next_pos`, so this diff will be 0, which I assume is not the intention (it will also affect the direction, thus speed calculated above).

Or perhaps I misunderstood the intent of this code?

Nemikolh replied:
> I assume next_pos will hold the updates coming from the server, and will try to do some prediction if non is received, correct?

That's correct.

> But after the first iteration, entity.pos is assigned to entity.next_pos

That's not what is done here. This is only done if the prediction would move the entity farther than the `next_pos`. Typical example being when the object is not moving.

Commit:
Added quick proxy and code for parsing messages from Lycan

Commit:
Added serialisation of commands, and separated mock code

Commit:
Start injecting LycanService at what seems the right place

Commit:
Switch to packed updates

Needs testing, and needs latest version of Lycan

Commit:
Move the LycanService into the runner module

Nemikolh committed:
Rename LycanMessage*. Remove fake commands.

Nemikolh committed:
Add GameInput, Refactor parts to be created by the canvas.

Nemikolh committed:
Fix travis.

Nemikolh committed:
Add ugly HUD and update entities positions

Nemikolh committed:
Update player position and fix bugs with actions updates.

Nemikolh committed:
Send always the last action saved.

Nemikolh committed:
Player position is predicted and proxy is now crashing on close errors.

Nemikolh committed:
Avoid making WebGL calls when it is not necessary.

Nemikolh committed:
Upgrade to typescript 2.1

Nemikolh committed:
Change server to use ImplicitAny false and assert keys in calls to pick.

This allowed us to find two bugs!

Nemikolh committed:
Add SpriteBatchObject to render multiple sprites.

Nemikolh committed:
Fix a rendering bug found by @Azamung.

Nemikolh committed:
Fix bugs where newly created maps and uploaded chipsets did not appear.

This seems to be more or less related to a previous update of angular.
Now we make sure to run in the Angular Zone every message received
from the server.

Nemikolh commented:
@lethom : The last commit should solve the problem in #50. I managed to reproduce the issue with a fresh install of dilia. Sorry for how it is solved. Once the runner is in a semi-ish stable state, I will make sure to get back to the old way of  submitting patches on dilia. I hope this is fine by you.

Nemikolh committed:
Add visual representation of entities. Fixed interpolation of positions.

lthms commented:
@Nemikolh It is more than okay. For now, our main concern should be fixing bug, not how we are fixing it. You’ve both done an amazing job.

lthms commented on editor/modules/aariba/behaviour.ts:
```diff
@@ -21,7 +21,7 @@ let SAFE_INSERT_BEFORE_TOKENS =
 // does not behave identically. Furthermore, I think we should
 // match the brackets. (Lookup for the next possible bracket).
-let context;
+let context: any;
```
You’ve got a lot of `any` in you code. Do you plan to (*can you*) be more specific in the future?

Nemikolh replied:
Oh nice catch! I think I will create an issue for this one.

To understand why we have so many `any`: basically the type definition that we have for the JavaScript Ace library is incomplete. When I wrote this code, I based it on an example written in JavaScript. For many of those function I had no idea what the type was. Once, I got ideas about what it could be, I couldn't find them in the type definition. So I got lazy and didn't annotate with a type all the arguments. However, with the `noImplicitAny` check from TypeScript that I have now turned on, the use of `any` must be explicit.

Nemikolh replied:
#57

lthms commented on editor/modules/runner/game-engine.ts:
```diff
@@ -0,0 +1,143 @@
+                });
+                handle.addLayer(layer);
+            }
+            // this.camera.centerOn(handle.build());
```
If you are leaving this line for a reason, you should add a sentence to explain why.

Nemikolh replied:
Good catch! I will remove it.

lthms replied:
It looks like you didn't. d:

Nemikolh replied:
oh indeed!

lthms commented on editor/modules/runner/game-engine.ts:
```diff
@@ -0,0 +1,143 @@
+        this.surface.setActivePipeline(pipeline);
+        this.surface.focus();
+
+        // TODO(Nemikolh): Move this elsewhere.
```
If merged as is, you should consider creating an issue.

lthms replied:
have you?

Nemikolh replied:
I think it's fine as is. I'm not sure exactly what I want to do about this code and I feel like creating an issue for it might be overkill.

lthms commented on editor/modules/runner/game-input.ts:
```diff
@@ -0,0 +1,140 @@
+        this.keyboard.register('azerty', Actions.UP, Keys.Z);
+        this.keyboard.register('azerty', Actions.DOWN, Keys.S);
+        this.keyboard.register('azerty', Actions.LEFT, Keys.Q);
+        this.keyboard.register('azerty', Actions.RIGHT, Keys.D);
```
Can we rely on some library for that? There are a lot more keyboard layout in the wild (for instance, bépo and dvorak d:)

Nemikolh replied:
That's a good point, I think we could use this one: https://github.com/ahtcx/layoutline. I'll remember to create an issue once this lands. We'll also need an interface and save the preference in a cookie.

Nemikolh replied:
#58

lthms commented on editor/modules/runner/game-state.ts:
```diff
@@ -0,0 +1,102 @@
+
+    entities: Entity[] = [];
+
+    // The player entity
```
I am not sure this comment adds any new information d:

Nemikolh replied:
XD

lthms commented on editor/modules/runner/game-state.ts:
```diff
@@ -0,0 +1,102 @@
+    ) {
+        this.lycan.getUpdateStream().subscribe(up => {
+            switch (up.kind) {
+                case 'NewEntity':
```
Is it possible to use an enum here?

Nemikolh replied:
This act exactly like Rust enum (sum type). The `kind` field act like a discriminant. Everything is type-checked. If you play with it, and replace `'NewEntity'` with `'newEntity'` you'll get an error. If you replace it with another valid variant you'll get an error within the case. :D

lthms replied:
Okay, great. The use of single quotes misled me.

lthms commented on editor/modules/runner/keyboard-layout.ts:
```diff
@@ -0,0 +1,162 @@
+// the keyboard layout of the user is QWERTY
+// or AZERTY.
+// We do so by assuming that we use ZQSD and WASD
+// to control the character.
```
same as before: there is no "standard way" to do that somewhere?

Nemikolh committed:
Fix compile error on server after rebase.

Azamung committed:
Fix vagrant on window 10

Avec Joan

lthms commented:
He did it!

Nemikolh committed:
Moved Proxy within dilia.

Nemikolh committed:
Augment cpu and RAM on VirtualBox VM.

Nemikolh committed:
Fix formatting.

Nemikolh committed:
Fix missing import. Convert `console.log` to `info`. Improve README.

Nemikolh committed:
Allow reconnection to lycan server. Provide errors infos to client.

Nemikolh committed:
Fix bug where null messages where being sent. Force reinit of connection

This is still work in progress. Once we have the console, it will
be easier to understand what is going on between lycan and dilia.

Nemikolh committed:
Address @lethom's comments.

Nemikolh committed:
Upgrade TSLint and friends. Update the code as appropriate.

Nemikolh commented:
@lethom, @Vaelden I hope you are fine with the merge. `master` was really starting to get out of sync and I would like to start two things in parallel (a UI rework and the console addition).

lthms commented:
@Nemikolh, at the end, you remain the master of `master`! You did well.

Nemikolh commented:
é_é
Thanks! :D
